File: src/controllers/SpellEditor.java
package controllers;

import application.ControllerInitilizer;
import javafx.fxml.FXML;
import javafx.scene.control.Button;
import javafx.scene.control.ComboBox;
import javafx.scene.control.TextField;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.input.KeyEvent;
import javafx.scene.input.MouseEvent;
import models.Spell;
import models.lists.Lists;
import models.lists.Spells;
import staticClasses.Listeners;

public class SpellEditor implements ControllerInitilizer
{
    private Spell      previouslySelectedSpell;
    private static int startSpell;
    private Spells     spells;

    @FXML
    void setSelectedSpell()
    {
	this.saveData();
	Spell spell = spellsC.getSelectionModel().getSelectedItem();
	Spell defaultSpell = spells.getDModel(spell);
	previouslySelectedSpell = spell;
	powerT.setText(Integer.toString(spell.getPower()));

	int defaultTargeting = defaultSpell.getTarget();
	if (defaultTargeting == 4)
	{
	    defaultTargeting = 3;
	}
	int targeting = spell.getTarget();
	if (targeting == 4)
	{
	    targeting = 3;
	}
	targetingC.getSelectionModel().select(targeting);
	costT.setText(Integer.toString(spell.getCost()));
	domainC.getSelectionModel().select(spell.getDomain());
	elementC.getSelectionModel().select(spell.getElement());
	unknown1T.setText(Integer.toString(spell.getUnknown1()));
	unknown2T.setText(Integer.toString(spell.getUnknown2()));

	defaultPowerT.setText(Integer.toString(defaultSpell.getPower()));
	defaultTargetingT.setText(targetingC.getItems().get(defaultTargeting));
	defaultCostT.setText(Integer.toString(defaultSpell.getCost()));
	defaultDomainT.setText(domainC.getItems().get(defaultSpell.getDomain()));
	defaultElementT.setText(elementC.getItems().get(defaultSpell.getElement()));
	defaultUnknown1T.setText(Integer.toString(defaultSpell.getUnknown1()));
	defaultUnknown2T.setText(Integer.toString(defaultSpell.getUnknown2()));
    }

    @FXML
    void spellDecrement(MouseEvent event)
    {
	spellsC.getSelectionModel().selectPrevious();
    }

    @FXML
    void spellIncrement(MouseEvent event)
    {
	spellsC.getSelectionModel().selectNext();
    }

    @FXML
    void removeNonNumbers(KeyEvent event)
    {
	Listeners.removeNonNumbers(event, null);
    }

    @Override
    public void saveData()
    {
	Spell s = previouslySelectedSpell;
	if (s == null) { return; }
	s.setPower(Integer.parseInt(powerT.getText()));

	int targeting = targetingC.getSelectionModel().getSelectedIndex();
	if (targeting == 3)
	{
	    targeting++;
	}
	s.setTarget(targeting);

	s.setCost(Integer.parseInt(costT.getText()));
	s.setDomain(domainC.getSelectionModel().getSelectedIndex());
	s.setElement(elementC.getSelectionModel().getSelectedIndex());
	s.setUnknown1(Integer.parseInt(unknown1T.getText()));
	s.setUnknown2(Integer.parseInt(unknown2T.getText()));

    }

    @FXML
    void initialize()
    {
	spells = new Spells();
	spells.addUsefulModels(Lists.getSpells(), false);
	spellsC.setItems(spells.getModels());
	this.addChangeListenerToTextFields();
	this.initializeComboboxes();
	spellsC.getSelectionModel().select(startSpell);
	this.setSelectedSpell();

	ImageView i = (ImageView) spellUp.getGraphic();
	i.setImage(new Image(this.getClass().getResourceAsStream("images/Triangle.png")));
	i = (ImageView) spellDown.getGraphic();
	i.setImage(new Image(this.getClass().getResourceAsStream("images/Triangle.png")));
    }

    private void addChangeListenerToTextFields()
    {
	powerT.focusedProperty().addListener(Listeners.textFieldFocusListener(65535));
	costT.focusedProperty().addListener(Listeners.textFieldFocusListener(65535));
	unknown1T.focusedProperty().addListener(Listeners.textFieldFocusListener(255));
	unknown2T.focusedProperty().addListener(Listeners.textFieldFocusListener(255));
    }

    private void initializeComboboxes()
    {
	targetingC.getItems().add("Single Enemy");
	targetingC.getItems().add("All Enemies");
	targetingC.getItems().add("Ally");
	targetingC.getItems().add("Map");

	domainC.getItems().add("All");
	domainC.getItems().add("Battle");
	domainC.getItems().add("Map");

	elementC.getItems().addAll(Spell.getElements());
    }

    @FXML
    private ComboBox<Spell>  spellsC;
    @FXML
    private TextField	     powerT;
    @FXML
    private ComboBox<String> targetingC;
    @FXML
    private TextField	     costT;
    @FXML
    private ComboBox<String> domainC;
    @FXML
    private ComboBox<String> elementC;
    @FXML
    private TextField	     unknown1T;

    @FXML
    private TextField	     unknown2T;

    @FXML
    private TextField	     defaultPowerT;
    @FXML
    private TextField	     defaultTargetingT;
    @FXML
    private TextField	     defaultCostT;
    @FXML
    private TextField	     defaultDomainT;
    @FXML
    private TextField	     defaultElementT;
    @FXML
    private TextField	     defaultUnknown1T;
    @FXML
    private TextField	     defaultUnknown2T;
    @FXML
    private Button	     spellUp;
    @FXML
    private Button	     spellDown;

    @Override
    public void saveState()
    {
	startSpell = spellsC.getSelectionModel().getSelectedIndex();
    }
}

File: src/application/controllers/SpellEditor.kt
package application.controllers

import application.ControllerInitilizer
import application.models.Spell
import application.models.lists.Lists
import application.models.lists.Spells
import application.staticClasses.Listeners
import javafx.collections.FXCollections
import javafx.fxml.FXML
import javafx.scene.control.Button
import javafx.scene.control.ComboBox
import javafx.scene.control.TextField
import javafx.scene.image.Image
import javafx.scene.image.ImageView
import javafx.scene.input.KeyEvent

class SpellEditor : ControllerInitilizer
{
	private var previouslySelectedSpell: Spell? = null
	private var spells = Spells(Lists.spells, false)

	@FXML fun setSelectedSpell()
	{
		this.saveData()
		val spell = spellsC.selectionModel.selectedItem
		val defaultSpell = spells.getDModel(spell)
		previouslySelectedSpell = spell
		powerT.text = Integer.toString(spell.power)

		var defaultTargeting = defaultSpell.target
		if (defaultTargeting == 4) defaultTargeting = 3

		var targeting = spell.target
		if (targeting == 4) targeting = 3
		targetingC.selectionModel.select(targeting)
		costT.text = Integer.toString(spell.cost)
		domainC.selectionModel.select(spell.domain)
		elementC.selectionModel.select(spell.element)
		unknown1T.text = Integer.toString(spell.unknown1)
		unknown2T.text = Integer.toString(spell.unknown2)

		defaultPowerT.text = Integer.toString(defaultSpell.power)
		defaultTargetingT.text = targetingC.items[defaultTargeting]
		defaultCostT.text = Integer.toString(defaultSpell.cost)
		defaultDomainT.text = domainC.items[defaultSpell.domain]
		defaultElementT.text = elementC.items[defaultSpell.element]
		defaultUnknown1T.text = Integer.toString(defaultSpell.unknown1)
		defaultUnknown2T.text = Integer.toString(defaultSpell.unknown2)
	}

	@FXML fun spellDecrement() = spellsC.selectionModel.selectPrevious()

	@FXML fun spellIncrement() = spellsC.selectionModel.selectNext()

	@FXML fun removeNonNumbers(event: KeyEvent) = Listeners.removeNonNumbers(event, null)

	override fun saveData()
	{
		val s = previouslySelectedSpell ?: return
		s.power = Integer.parseInt(powerT.text)

		var targeting = targetingC.selectionModel.selectedIndex
		if (targeting == 3)
		{
			targeting++
		}
		s.target = targeting

		s.cost = Integer.parseInt(costT.text)
		s.domain = domainC.selectionModel.selectedIndex
		s.element = elementC.selectionModel.selectedIndex
		s.unknown1 = Integer.parseInt(unknown1T.text)
		s.unknown2 = Integer.parseInt(unknown2T.text)

	}

	@FXML fun initialize()
	{
		spells.chronologicalIndexSort()
		spellsC.items = spells.models
		addChangeListenerToTextFields()
		initializeComboboxes()
		spellsC.selectionModel.select(startSpell)
		this.setSelectedSpell()

		var i = spellUp.graphic as ImageView
		i.image = Image(this.javaClass.getResourceAsStream("images/Triangle.png"))
		i = spellDown.graphic as ImageView
		i.image = Image(this.javaClass.getResourceAsStream("images/Triangle.png"))
	}

	private fun addChangeListenerToTextFields()
	{
		powerT.focusedProperty().addListener(Listeners.textFieldFocusListener(65535))
		costT.focusedProperty().addListener(Listeners.textFieldFocusListener(65535))
		unknown1T.focusedProperty().addListener(Listeners.textFieldFocusListener(255))
		unknown2T.focusedProperty().addListener(Listeners.textFieldFocusListener(255))
	}

	private fun initializeComboboxes()
	{
		targetingC.items = FXCollections.observableArrayList("Single Enemy", "All Enemies", "Ally", "Map")
		domainC.items = FXCollections.observableArrayList("All", "Battle", "Map")
		elementC.items.addAll(Spell.elementText)
	}

	@FXML lateinit var spellsC: ComboBox<Spell>
	@FXML lateinit var powerT: TextField
	@FXML lateinit var targetingC: ComboBox<String>
	@FXML lateinit var costT: TextField
	@FXML lateinit var domainC: ComboBox<String>
	@FXML lateinit var elementC: ComboBox<String>
	@FXML lateinit var unknown1T: TextField
	@FXML lateinit var unknown2T: TextField
	@FXML lateinit var defaultPowerT: TextField
	@FXML lateinit var defaultTargetingT: TextField
	@FXML lateinit var defaultCostT: TextField
	@FXML lateinit var defaultDomainT: TextField
	@FXML lateinit var defaultElementT: TextField
	@FXML lateinit var defaultUnknown1T: TextField
	@FXML lateinit var defaultUnknown2T: TextField
	@FXML lateinit var spellUp: Button
	@FXML lateinit var spellDown: Button

	override fun saveState()
	{
		startSpell = spellsC.selectionModel.selectedIndex
	}

	companion object
	{
		private var startSpell: Int = 0
	}
}
